### PR TITLE
More native flake support

### PR DESCRIPTION
**Pull request by Gerschtli**
This is my first draft of a more native integration of flakes support including

* conventions for `nixOnDroidConfigurations` and `nixOnDroidModules` flake outputs
* support for flakes in `nix-on-droid` tool
* support for `extraModules` and `extraSpecialArgs` similar to `home-manager` and `NixOS`

Any feedback on the API and documentation would be appreciated. The actual code changes are not tested yet, but the earlier I get feedback on my main ideas the better :)

This one supersedes https://github.com/t184256/nix-on-droid/pull/145, incorporating all its changes. The reason to not wait for its merge is to only change the `lib` output of the `nix-on-droid` once and I wanted these fields like `extraModules` and `extraSpecialArgs` in there :)

Ping @zhaofengli @ShamrockLee @bbigras as you were involved in flake related stuff lately in nix-on-droid, maybe you want to have a look at that too, to finish the first experimental native flake support.

### CODE REVIEW

**Comment by t184256** [comment]
> This one supersedes #145, incorporating all its changes. The reason to not wait for its merge is to only change the lib output of the nix-on-droid once and I wanted these fields like extraModules and extraSpecialArgs in there :)

Uh, sorry for creating additional work for you, but I'd like #145 merged separately with @zhaofengli's commit being distinct, for authorship tracking purposes. Please rebase.

**Comment by Gerschtli** [comment]
Thats alright for me, I will do after I did my testing.

**Comment by Gerschtli** [comment]
Everything is rebased and tested with `nix-on-droid on-device-test`. If I missed a test case, I will add it :)

So ready for final review!

**Comment by t184256** [comment]
Do you intend it to go into release-21.11 or to 'stage' in master following the release?

**Comment by Gerschtli** [comment]
That is debatable I would say. Because flake stuff is experimental, breaking changes are expected to appear anytime. And as this PR does not break any behaviour regarding the default way of things and keeping the branches release-21.11 and master in sync brings more pain than benefit when putting this PR only into master.

So in the interest of maintainability and flake support still being experimental, I would suggest going straight to release-21.11.

File: tests/on-device/config-h-m.bats
Comment (t184256) [inline tests/on-device/config-h-m.bats:12]: Why both?

**Review by t184256 — COMMENTED** [review]
Looks less scary now when reviewed commit-by-commit. The "what do we revert to in tests" is my biggest question mark at the moment.

File: modules/default.nix
Comment (Gerschtli) [reply modules/default.nix:7]: Ah thank you for mentioning that. I tested it locally and was a bit confused by the notation `<home-manager>` and how this should resolve to a path without any other function call. If `<home-manager>` works (I will test it with on-device-test), I will change it back.

File: nix-on-droid/default.nix
Comment (Gerschtli) [reply nix-on-droid/default.nix:3]: Since we are in release 21.11, we should be able to assume nixpkgs 21.11. Therefore using the stable nix 2.4 release instead of an moving pointer to a nix release that supports flakes felt more reliable.

File: nix-on-droid/default.nix
Comment (Gerschtli) [reply nix-on-droid/default.nix:18]: Didn't think of that, I will have a look at the closure sizes. Afterwards we can decide :)

File: nix-on-droid/nix-on-droid.sh
Comment (Gerschtli) [reply nix-on-droid/nix-on-droid.sh:138]: I 100% agree, will put a comment in there :) Sometimes you are so deeply in the code that this feels natural at this moment :D

File: README.md
Comment (Gerschtli) [reply README.md:198]: Ah thank you!

File: tests/on-device/lib.bash
Comment (Gerschtli) [reply tests/on-device/lib.bash:27]: Not exactly. If you are developing on the nix-on-droid script, this line will revert everything and go back to the initial state. switch_to_default_config will then activate the latest version of nix-on-droid you want to test and only then you can start testing.

File: tests/on-device/.run.sh
Comment (Gerschtli) [reply tests/on-device/.run.sh:28]: Ah damn it, that was my "initial" generation. Wasn't meant to be committed. Will change it to `$PROFILE_DIRECTORY/activate`

File: tests/on-device/config-h-m.bats
Comment (Gerschtli) [reply tests/on-device/config-h-m.bats:33]: Yeah, it doesn't hurt, you are right

File: tests/on-device/gnumake.bats
Comment (Gerschtli) [reply tests/on-device/gnumake.bats:9]: Okay, will revert it

File: tests/on-device/config-h-m.bats
Comment (Gerschtli) [reply tests/on-device/config-h-m.bats:12]: `nix-channel --remove home-manager` removes link in `~/.nix-defexpr`, so that the NIX_PATH is predictable again.

`mv ~/.nix-channels.bak ~/.nix-channels` just ensures that other channels configured before are restored, so that your installation is not messed up and can be used normally afterwards.

File: README.md
Comment (bbigras) [inline README.md:174]:
```suggestion
    nixpkgs.url = "nixpkgs/release-21.11";
```

use the registry :)

File: README.md
Comment (bbigras) [inline README.md:198]:
```suggestion
    home-manager.url = "home-manager/release-21.11";
```

File: tests/on-device/config-flake-h-m.flake.nix
Comment (bbigras) [inline tests/on-device/config-flake-h-m.flake.nix:5]:
```suggestion
    nixpkgs.url = "nixpkgs/release-21.11";
```

File: tests/on-device/config-flake-h-m.flake.nix
Comment (bbigras) [inline tests/on-device/config-flake-h-m.flake.nix:6]:
```suggestion
    home-manager.url = "home-manager/release-21.11";
```

**Comment by Gerschtli** [comment]
Thank you for your review!

The idea of state reversion is: Change everything back to the original state (like removing files, removing nix channels). Before each test there is the activation of the "initial" generation which marks a point in time before starting development (aka some stable release of nix-on-droid) where nothing is broken. Then you switch to the default config with the latest nix-on-droid channel version to have the latest version, that you developing on, installed and activated on your system. Then the test starts.

File: README.md
Comment (bbigras) [reply README.md:174]: > If I configure my system registry to let it point nixpkgs to my local revision (a path to a forked repo on my machine), will that be picked up?

I think so if you use `nix registry pin`. Otherwise, I don't see the point of pinning a flake.

> What advantages do we have when using this shorthands?

imho, it's cleaner because it's shorter, and I would tend to trust that the registry is pointing to the right place.

Comment (bbigras) [reply README.md:174]: also if every flake would use the registry, it would be a lot easier to pin nixpkgs to unstable or something and have all your projects follow it. Like in my case I use flakes for many projects, I think I'll start pinning nixpkgs to unstable.

File: tests/on-device/config-flake-h-m.bats
Comment (t184256) [reply tests/on-device/config-flake-h-m.bats:8]: If the test is testing "X was there, now it's not" or vice versa, but the first check movest to some other file, it's prone to changing with time and then suddenly we can find ourselves not testing that activation changed something.

Comment (Gerschtli) [reply tests/on-device/config-flake-h-m.bats:8]: Okay understood. So my idea would then be: move switch to default without assertions to setup because it is just setup. Add all for this specific test case relevant assertions at the top of the test and remove all other ones at the top that do not bring value for this specific test. I will mark these assertions with a comment.

File: README.md
Comment (Gerschtli) [reply README.md:174]: Because these flakes here should show best practices and a good starting point, I prefer to not use shorthand registry identifier.

If you pin nixpkgs on your system to any revision and want to use it for your personal projects, you can do so. If you are just copying these examples with the registry urls, you may use a pinned nixpkgs without knowing. I recently had a problem with a pinned nixpkgs and was searching quite a while before realising that `nixpkgs` input in my flake was referencing a local store path instead of a nixpkgs git revision, which my CI obviously could not find.

So feel free to use registry shortcuts if you know what you are doing, but I would rather not recommend it as a best practice here because it can be tedious to debug.

Comment (t184256) [reply README.md:174]: Huh? Registry doesn't pin to commits. Personally, I'd trust the nixpkgs entry value in the registry more than the contents of nixpkgs itself.

Comment (bbigras) [reply README.md:174]: yeah. pins must be local. in `~/.config/nix/registry.json` or something.

File: nix-on-droid/default.nix
Comment (Gerschtli) [reply nix-on-droid/default.nix:18]:
```
[nix-shell:~]$ nix path-info nixpkgs.nix_2_4 -r --json | jq 'map(.narSize) | add'
139268464

[nix-shell:~]$ nix path-info nixpkgs.nix -r --json | jq 'map(.narSize) | add'
132446688

[nix-shell:~]$ nix path-info nixpkgs.nix nixpkgs.nix_2_4 -r --json | jq 'map(.narSize) | add'
150507760
```

So the difference is 150507760 - 132446688 = 18061072 B = 18.06 MiB if I am not wrong. Sounds okay for me.

File: README.md
Comment (Gerschtli) [reply README.md:174]: You can pin a special revision or a path to a local forked nixpkgs with a registry which then pins to commits. The one thing I trust the most not to have to trust any registry being correct. But as I said: Everyone can use it as he/she likes. This only tries to show a reasonable default that does not need any knowledge about the nix registry. I can add a small paragraph mentioning that it is possible to use registry shortcuts but that would just be a duplication of the nix manual.

File: nix-on-droid/default.nix
Comment (t184256) [reply nix-on-droid/default.nix:18]: That's great, I expected worse. Thanks for checking.

File: README.md
Comment (t184256) [reply README.md:174]: FWIW, I'm fine with not using registry.

**Comment by Gerschtli** [comment]
@t184256 These are a lot of fixup commits to keep reviewing for you all easy. After approval I will autosquash this branch and rebase it on latest master. So please keep that in mind before merging it :)

I think I addressed everything, PTAL and resolve (or unresolve) your comments if possible :)

File: nix-on-droid/default.nix
Comment (t184256) [reply nix-on-droid/default.nix:3]: Fine as long as we don't forget to unpin it next time, I guess. And I hope next time we'll revisit it anyway because we'll have flakes in stable.

File: tests/on-device/lib.bash
Comment (t184256) [reply tests/on-device/lib.bash:27]: OK. I propose `restore to initial generation before the start of each test` or `restore to pre-testing generation before the start of each test` then.

File: nix-on-droid/default.nix
Comment (Gerschtli) [reply nix-on-droid/default.nix:3]: I think for 21.05 nix 2.4 will become the default in nixpkgs and flakes should be marked as stable. After that we will remove that here.

File: tests/on-device/lib.bash
Comment (Gerschtli) [reply tests/on-device/lib.bash:27]: Not sure if you saw my rewording `restore an generation of a stable release before start testing` already. Is that fine? Otherwise I'll  pick one of yours.

Comment (t184256) [reply tests/on-device/lib.bash:27]: yeah, I was commenting on the new one from the commit view. `a`/`and` + `start`/`starting`  + unclear what `stable release` means when the reverting point is fully arbitrary now as you've switched it, philosophically, from testing default <-> different confs to "state under test" <-> different confs. Not that there's anything wrong with it.

Comment (Gerschtli) [reply tests/on-device/lib.bash:27]: Agree, I like your second wording more than mine, will change it :)

**Comment by t184256** [comment]
I don't see any more stuff to complain about with my naked eye. Both flake tests failed with `attribute 'nixOnDroidConfigurations' missing` when tested on device. Please rebase, you can also squash freely up to 9caebbe inclusive.

**Comment by Gerschtli** [comment]
Alright. I am a bit confused because these tests ran perfectly fine in your fakedroid setup. Will have a look at that.

I would rather not squash the commits to have a clean history of smaller commits. Will rebase in the evening.

**Comment by t184256** [comment]
squash the fixups, I meant

**Comment by Gerschtli** [comment]
Autosquashed and rebased, no actual changes done. Will now have a look at your mentioned problem.

**Comment by Gerschtli** [comment]
Could not be reproduced via fakedroid, on a real device or the CI (CI is failing because it can not push to cachix).

Could you verify again if the tests are still failing for you?

**Comment by Gerschtli** [comment]
Okay :+1: 

Yeah, I always forget that I have write access to your repo (thanks for that BTW :) ). I'm used to work on my forks :D

**Comment by t184256** [comment]
Is it stripping review on each rebase?

**Comment by Gerschtli** [comment]
No, not that I know of, I just requested a review from you again :)

**Review by t184256 — APPROVED** [review]

**Pull request merged**
